### ISSUE TITLE
Race condition in Order::setInvoiceDetails

### ISSUE THREAD

**Issue by neonVoice** [PrestaShop/PrestaShop#23356]
#### Describe the bug

Hello, I'm coming from MSSQL background, so please dont harsh on me for using word **transaction** a lot, im using this project partially as hobby/work and not proficient in PHP/MySQL.

I am using payment module from my local bank organization(stands as an internet acquiring ith a lot of payment options) which change order statuses based on bank response and according to their scheme of message exchange when an order is payed through for example VISA card or Google/Apple Pay it first gets checked on their end and then i get 2 notifications which are then sent to module front controller. There is exactly 2 statuses which they send - `AUTHORIZED`, and immediately after it `CONFIRMED`. `AUTORIZED` operation changes order status to `_PS_OS_PREPARATION_` and `CONFIRMED` changes it to `_PS_OS_PAYMENT_` respectively. According to log files which grab input data in front controller - this 2 notifications are sent very fast (without delay - basicaly 2 events fired at their end one after another) which both use `Order::setCurrentState` method causing high risk of **race conditions** (or more precisely - DIRTY WRITES when second transaction can't relate on data stored in database because, according to code logic second transaction should'nt even entered this code cave if first transaction would finish executing in time - basicaly data from first transaction isn't in DB yet).
TLDR:

1. I got my statuses in wrong order inside `ps_order_history` table. (_PS_OS_PAYMENT_) comes before (_PS_OS_PREPARATION_)
2. My payments are doubled(quadrupled) Instead of 1 invoice = 1 completed payment now i got 2 invoices and 4 payments. 1 invoice is closed with 1 payment, 3 other payments are hanging resulting in message: `Warning 6 684,20 ₽ paid instead of 1 671,05 ₽`
3. Invoices are doubled (latter invoice contains all data about order when i click on invoice and get PDF to open it in browser, former invoice displays an empty table with only subtotal fields filled)
4. Delivery Slips are doubled (same behavior like in point 3).

Code digging led me to some pieces of code. Further goes some kind of manual stack tracing without IDE and thoughts about what happened in terms of timeline.

(int) _PS_OS_PREPARATION_ = 3
(int) _PS_OS_PAYMENT_ = 2

```php
Order::setCurrentState($newOrderStatus)
```
this line gets called consequently for 2 states right one after another from 2 different threads.
```php
$use_existings_payment = !$this->hasInvoice();
```

this line returns `false` for both executions

```php
$history->changeIdOrderState(2, $this, false);
$history->changeIdOrderState(3, $this, false);
```
```php
if ($new_os->invoice && !$order->invoice_number) {
    $order->setInvoice($use_existing_payment);
} elseif ($new_os->delivery && !$order->delivery_number) {
    $order->setDeliverySlip();
}
```
not sure which branch executed here, but doesn't matter because at this moment of time both `!$order->invoice_number` and `!$order->delivery_number` checks pass and `setInvoice` or `setDeliverySlip` methods would get called which both later in the code execute this:
```php
$this->setInvoiceDetails($order_invoice);
```
however most likely method `setInvoice` get's called exactly in this situation because according to this block of code:
```php
if ($id = (int) $this->getOrderInvoiceIdIfHasDelivery()) {
    $order_invoice = new OrderInvoice($id);
} else {
    $order_invoice = new OrderInvoice();
}
```
in both scenarious new OrderInvoice **without $id** gets executed. As far as I'm concerned - after calling `$this->setInvoiceDetails($order_invoice);` exactly this method call results in insertion to table `ps_order_invoice` in this call to object model:
```php
$order_invoice->save();
```
Having 2 identical rows in table `ps_order_invoice` with all data filled led me to thinking that this line of code got executed at first glance.

In the world of MSSQL i would place code which ensures only 1 entity in DB exists at any given time in `transaction block` or `locks`, or if we are talking from DB perspective check before insertion:

```sql
INSERT table t
SELECT
  delivery_number = td.delivery_number,
  ...
FROM #temp_data td 
WHERE NOT EXISTS (
  SELECT *
  FROM table t
  WHERE td.id_order = t.id_order
    AND td.delivery_number IS NOT NULL
)
```

So, basically my question - is it possible to incorporate some kind of `lock or queue mechanism` to handle similar kind of stuff from user code? Again, im not that experienced in PHP know nothing about ptreads library and so on. For example i would want latter thread to wait on `$order` object after acquiring it by id and only execute code after former code got executed (status got changed in DB)

#### Expected behavior

Expected 1 invoice covered with 1 payment and 1 delivery slip

#### Steps to Reproduce

Hard to recreate artificially if we are talking about full chain of actions: need to call method Order::setOrderState one after another, preferable from 2 different requests, for example using postman or curl or some kind of script.

**Screenshots**

![statuses](https://user-images.githubusercontent.com/21147332/108775510-345a8080-7572-11eb-9c47-d5de224916ca.png)
![documents](https://user-images.githubusercontent.com/21147332/108775518-36bcda80-7572-11eb-9a00-b4f31981e273.png)
[invoice - 000061.pdf](https://github.com/PrestaShop/PrestaShop/files/6025401/invoice.-.000061.pdf)
[invoice - 000060.pdf](https://github.com/PrestaShop/PrestaShop/files/6025404/invoice.-.000060.pdf)
[slip 61 - 000061.pdf](https://github.com/PrestaShop/PrestaShop/files/6025405/slip.61.-.000061.pdf)
[slip 60 - 000061.pdf](https://github.com/PrestaShop/PrestaShop/files/6025406/slip.60.-.000061.pdf)

Clicking on delivery slip 000060 in order to get PDF resulted in filename called 000061 but the content was still from invoice 60.

#### Additional information

* PrestaShop version: 1.7.7.1
* PHP version: 7.2.19

**Comment by matks** [comment]
> #### Describe the bug
> Hello, I'm coming from MSSQL background, so please dont harsh on me for using word **transaction** a lot, im using this project partially as hobby/work and not proficient in PHP/MySQL.

Hi @neonVoice and thanks for this long bug report, thank you for taking the time to write this 👍 

There's no issue talking about transactions 😄 , actually we [wish we could rely on them](https://github.com/PrestaShop/PrestaShop/pull/21740) too but for a long time PrestaShop supported both MyISAM and InnoDB. We are considering dropping MyISAM which would allow us to start using transactions, but the decision has not been acted yet.

This is why we have multiple issures related to DIRTY writes, race conditions and so on ; so your bug report makes sense.

**Comment by hibatallahAouadni** [comment]
Hello @neonVoice 

Thanks for the detailed bug report!
> Hard to recreate artificially if we are talking about full chain of actions: need to call method Order::setOrderState one after another, preferable from 2 different requests, for example using postman or curl or some kind of script.

So, to reproduce the issue we need to call this method twice **`Order::setOrderState`** simultaneously (or one after another) from a different WS and set the same status for the same Order, is that so?

Thanks!

**Comment by neonVoice** [comment]
> 
> 
> Hello @neonVoice
> 
> Thanks for the detailed bug report!
> 
> > Hard to recreate artificially if we are talking about full chain of actions: need to call method Order::setOrderState one after another, preferable from 2 different requests, for example using postman or curl or some kind of script.
> 
> So, to reproduce the issue we need to call this method twice **`Order::setOrderState`** simultaneously (or one after another) from a different WS and set the same status for the same Order, is that so?
> 
> Thanks!

Hello @hibatallahAouadni, consider this front controller code located in file `notification.php`:
```php
class TinkoffNotificationModuleFrontController extends ModuleFrontController
{
    public function postProcess()
    {
        try {
            Tools::safePostVars();
            $order = getOrderById(Tools::getValue('OrderId'));
            
            $newOrderStatus = _PS_OS_CANCELED_;

            switch (Tools::getValue('Status')) {
                case 'AUTHORIZED':
                    $newOrderStatus = _PS_OS_PREPARATION_;
                    break;
                case 'CONFIRMED':
                    $newOrderStatus = _PS_OS_PAYMENT_;
                    break;
                case 'REFUNDED':
                    $newOrderStatus = _PS_OS_REFUND_;
                    break;
            }

            $order->setCurrentState($newOrderStatus);
            die('OK');
        } catch (Exception $e) {
            die('NOTOK');
        }
    }
    
    private function getOrderById($id_order)
    {
        $order = Db::getInstance()->getRow(
            'SELECT * FROM `'._DB_PREFIX_.'orders`
			WHERE `id_order` = '.(int)$id_order
        );

        return ($order['id_order'] > 0) ? new Order($order['id_order']) : null;
    }
}
```

it is accessible within url: https://xxx.yyy/module/tinkoff/notification

We need to perform 2 consecutive POST requests to this url like this:
```json
{
  "OrderId": 100,
  "Status": "AUTHORIZED"
}
```
```json
{
  "OrderId": 100,
  "Status": "CONFIRMED"
}
```

If we're talking about time perspective - they are executed one after another in the same code with difference like 200 microseconds, like this (example written in fast python):

```python3
import requests

url = "http://xxx.yyy/module/tinkoff/notification"
headers = {'Content-type': 'application/json', 'Accept': 'text/plain'}
data = {"OrderId":100, "Status":"AUTHORIZED"}
r = requests.post(url, data=json.dumps(data), headers=headers)
data = {"OrderId":100, "Status":"CONFIRMED"}
r = requests.post(url, data=json.dumps(data), headers=headers)
```

So, answering your questions:
`So, to reproduce the issue we need to call this method twice Order::setOrderState simultaneously (or one after another) from a different WS and set the same status for the same Order, is that so?`
1. Call this method twice `Order::setOrderState` simultaneously (or one after another) ? - Twice, one after another
2. From a different WS ? - Nope, one WS is enough, but it could be performed simultaneously if requests send different statuses.
3. Set the same status for the same Order ? - Requests set different statuses for the same order.

**Comment by hibatallahAouadni** [comment]
Hello @neonVoice 

Well, I followed these steps:

1. Create a WS to call this method **`Order::setOrderState`** twice

- Request URL: **`http://your_shop_url//api/order_histories`**
- Method: **`POST`**
- Request body: 

```
<prestashop>
    <order_history>
    	<id_employee>1</id_employee>
    	<id_order_state>2</id_order_state>
    	<id_order>29</id_order>
    	<date_add></date_add>
    </order_history>
</prestashop>
```

2. Launch this WS twice, one after another (in the second attempt I change **`id_order_state`** to **`3`** then to **`5`**
3. It wroks perfectly without any error => The order status has been changed in BO
4. I only have one invoice and one delivery slip

![image](https://user-images.githubusercontent.com/23308427/109084480-2474a500-7708-11eb-9242-6161733c77b6.png)

![image](https://user-images.githubusercontent.com/23308427/109084495-29d1ef80-7708-11eb-950e-8a2cb70302a9.png)

Please check and feedback.
Thanks!

**Comment by neonVoice** [comment]
Unfortunately i wasn't able to replicate this bug either both on my test and production environments. All i can say is that when order was generated - some of the payment module code was overriden in `override` folder for logging purposes - i was just logging all incoming responses, then calling parent method.
For example class declaration in `/modules/xxx/controllers/front/notification.php`:
```php
class TinkoffNotificationModuleFrontController extends ModuleFrontController
```
while in folder `/override/modules/xxx/controllers/front/notification.php`
```php
class TinkoffNotificationModuleFrontControllerOverride extends TinkoffNotificationModuleFrontController
```
Stating that `TinkoffNotificationModuleFrontControllerOverride` extends `TinkoffNotificationModuleFrontController` i decided to do something like that:
```php
class TinkoffNotificationModuleFrontControllerOverride extends TinkoffNotificationModuleFrontController
{
    /**
     * @see FrontController::postProcess()
     */
    public function postProcess()
    {
        $data = [
            "get" => $_GET,
            "post" => $_POST
        ];
        $this->logs($data, "NOTIFICATION");
        
        parent::postProcess();
    }
    
    private function logs($requestData, $request)
    {
        // log send
        $log = '[' . date('D M d H:i:s Y', time()) . '] ';
        $log .= json_encode($requestData, JSON_UNESCAPED_UNICODE);
        $log .= "\n";
        file_put_contents(dirname(__FILE__) . "/tinkoff.log", $log, FILE_APPEND);

        $log = '[' . date('D M d H:i:s Y', time()) . '] ';
        $log .= $request;
        $log .= "\n";
        file_put_contents(dirname(__FILE__) . "/tinkoff.log", $log, FILE_APPEND);
    }
}
```
Later i moved all code from override directly to module folder.

**tldr: cant replicate**

**Comment by hibatallahAouadni** [comment]
Hello @neonVoice 

Well, you cnat' reproduce the issue anymore :tada: glad to hear.
I close this issue as **`can't reproduce`** but feel free to notify me or open a new issue if you reproduce it again :wink: 

Thanks!
